Make Reek more resilient on incomprehensible source.

troessner opened the pull request:
Fixes #995, heavily wip.

troessner commented:
This one is depending on #997

mvz commented:
#997 has been merged, so this can be rebased I think.

troessner commented:
Ready for review!

mvz commented:
I'll have a look tomorrow.

mvz commented on spec/reek/examiner_spec.rb:
```diff
@@ -108,5 +108,42 @@
+    context 'with an incomprehensible source that causes Reek to crash' do
+      let(:subject) do
+        examiner = described_class.new 'class C; def does_crash_reek; end; end'
+        allow(examiner).to receive(:smell_repository) do
```
I'm don't like it if the object under test is stubbed. Can we avoid this by either injecting `SmellRepository`, or stubbing `SmellRepository.new` directly?

troessner replied:
>> Can we avoid this by either injecting SmellRepository

I'd prefer that as well but our current Examiner constructor signature is not supporting so we'd have to add this. Which would mean that it would take 4 arguments, which I find quite a lot but still manageable. Are you suggesting to update the Examiner constructor to be like this:

```Ruby

    def initialize(source,
                   filter_by_smells: [],
                   configuration: Configuration::AppConfiguration.default,
                   smell_repository: Smells::SmellRepository.new)
end
```

?

>> or stubbing SmellRepository.new directly?

I can't see the benefit. For me this would be even worse since it is more opaque - you're stubbing something that is not even called directly thus making it harder to build up a mental image. Why do you think this is better?

mvz replied:
> Are you suggesting to update the Examiner constructor to be like this:

We probably need to inject the class instead of an instance, but yes, something like that.

> > or stubbing SmellRepository.new directly?
> I can't see the benefit. For me this would be even worse since it is more opaque - you're stubbing something that is not even called directly thus making it harder to build up a mental image. Why do you think this is better?

I think `SmellRepository.new` is actually called directly in `Examiner`. But to answer more broadly, I have found again and again that stubbing the object under test is really the most dangerous option, since it tends to hide bugs in the implementation of the object itself.

mvz commented on spec/reek/examiner_spec.rb:
```diff
@@ -108,5 +108,42 @@
     end
+
+    context 'with an incomprehensible source that causes Reek to crash' do
+      let(:subject) do
```
Can you replace this with either `subject do ...`, or (preferred) `let(:examiner) do`?

troessner replied:
Good point, fixed!

mvz commented on spec/spec_helper.rb:
```diff
@@ -3,6 +3,7 @@
 require_relative '../lib/reek/spec'
 require_relative '../lib/reek/ast/ast_node_class_map'
 require_relative '../lib/reek/configuration/app_configuration'
+require 'active_support/core_ext/string/strip'
```
Why did you move this?

troessner replied:
The overall rationale here was that we use `strip_heredoc` in multiple places so it would make sense to require this in the spec_helper.
But you're right, this really doesn't belong to this PR, I just undid this change.

mvz commented:
Just a few nits, then it's ready to go.

Drenmi commented on lib/reek/examiner.rb:
```diff
@@ -12,6 +12,19 @@ module Reek
+      This is most likely a Reek bug.
+      It would be great if you could report this back to the Reek team
+      by opening up a corresponding issue at https://github.com/troessner/reek/issues
+      Make sure to include the source in question, the Reek version
```
Just a thought. Would it be feasible to have all this information ready to be copied and pasted from the error message, rather than having the user go and gather it?

troessner replied:
Awesome idea!
I'll add this, thanks!

mvz commented on spec/reek/examiner_spec.rb:
```diff
@@ -108,5 +108,42 @@
+    context 'with an incomprehensible source that causes Reek to crash' do
+      subject do
+        examiner = described_class.new 'class C; def does_crash_reek; end; end'
+        allow(examiner).to receive(:smell_repository) do
```
(comment copied so it is not hidden)

> Are you suggesting to update the Examiner constructor to be like this:

We probably need to inject the class instead of an instance, but yes, something like that.

> >  or stubbing SmellRepository.new directly?

>  I can't see the benefit. For me this would be even worse since it is more opaque - you're stubbing something that is not even called directly thus making it harder to build up a mental image. Why do you think this is better?

I think SmellRepository.new is actually called directly in Examiner. But to answer more broadly, I have found again and again that stubbing the object under test is really the most dangerous option, since it tends to hide bugs in the implementation of the object itself.

troessner commented:
@mvz addressed of all your feedback in 2 separate commits, the first one making SmellRepository injectable and the second one making the specs less brittle.

I'll address 

>> Would it be feasible to have all this information ready to be copied and pasted from the error message, rather than having the user go and gather it?

in a separate pull request to not block this one (see #1019)

LGTM?

mvz commented:
It looks like you're not using the new dependency injection possibility in the specs?

troessner commented:
>> It looks like you're not using the new dependency injection possibility in the specs?

Exactly! I started out with adding it but when it came to the spec itself I realized that RSpec's awesome 

```Ruby
class_double('Reek::Smells::SmellRepository').as_stubbed_const
```

works as well for this very spec. I still left the DI in there because a) I think we should start using DI way more in our codebase and b) starting to DI SmellRepository might be useful for subsequent specs. That's why it's 2 separate commits.

I can also remove the DI for SmellRepository but I think it makes sense to keep it for above reasons.

WDYT?

mvz commented:
Ok, clear.
